cidc_api/utils/migrations.py

```diff
@@ -42,39 +43,67 @@ def run_all(self):
         for i, task in enumerate(self.tasks):
             try:
                 task.do()
+                self.done.add(i)
             except:
-                for done_task in self.tasks[:i]:
-                    done_task.undo()
+                self.rollback()
                 raise
 
+    def rollback(self):
+        """
+        Undo any work that has been carried out.
+        """
+        for i, task in enumerate(self.tasks):
+            if i in self.done:
+                task.undo()
+
 
 @contextmanager
 def migration_session():
     session = Session(bind=op.get_bind())
+    task_queue = RollbackableQueue()
 
     try:
-        yield session
+        yield session, task_queue
+        session.commit()
     except:
         session.rollback()
+        if task_queue:
+            task_queue.rollback()
```

Comment:
in case this `.rollback` call fails/raises, will we get both exceptions in the logs ?